internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,299 @@
+		return reconcile.Result{}, r.updateNADRWithURL(ctx, &veleroDR, req)
+	} else if !apierrors.IsNotFound(err) {
+		// some other errors, requeue to retry get
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, r.processDownloadRequest(ctx, &veleroDR, req)
+}
+
+// Delete expired download request
+// find associated velero downloadrequest and delete that first
+func (r *NonAdminDownloadRequestReconciler) deleteExpiredDownloadRequest(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("Deleting expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), req.Namespace)
+	if err := r.Delete(ctx, veleroDR); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), err)
+			// other errors, requeue to retry delete
+			return err
+		}
+	}
+	logger.V(1).Info("Deleting expired NonAdminDownloadRequest", req.Name, req.Namespace)
+	if err := r.Delete(ctx, req); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest", req.Name, err)
+			// other errors, requeue to retry delete
+			return err
+		}
+		// not found, stop requeue
+	}
+	return nil
+}
+
+// if url is available and not expired, then set status to completed
+func (r *NonAdminDownloadRequestReconciler) updateNADRWithURL(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	if veleroDR.Status.DownloadURL != constant.EmptyString &&
+		veleroDR.Status.Phase == velerov1.DownloadRequestPhaseProcessed &&
+		!veleroDR.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+		// copy req, prepare to patch
+		prePatch := req.DeepCopy()
+		// copy status to NADR from VDR
+		req.Status.VeleroDownloadRequest.Status = &veleroDR.Status
+		req.Status.Phase = nacv1alpha1.NonAdminPhaseCreated
+		// veleroDR is processed, update NADR status
+		// clear conditions
+		req.Status.Conditions = []metav1.Condition{}
+		if patchErr := r.Status().Patch(ctx, req, client.MergeFrom(prePatch)); patchErr != nil {
+			logger.Error(patchErr, "unable to patch status", req.Kind, req.Name)
+			return patchErr
+		}
+	}
+	return nil
+}
+
+// processDownloadRequest creates download requests after performing NaBSL checks on NAB/NAR from the downloadRequest
+func (r *NonAdminDownloadRequestReconciler) processDownloadRequest(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	var nab nacv1alpha1.NonAdminBackup // holds nonadminbackup
+	var nabName string                 // holds nonadminbackup name
+	switch req.Spec.Target.Kind {
+	case velerov1.DownloadTargetKindBackupLog,
```

Comment:
Are we allowing all kinds of download Request target ?
if not then we should definitely add a validation function that gets executed before the processDR func.

Reply:
yes. all.